library/pem.c

```diff
@@ -241,6 +241,28 @@ static int pem_aes_decrypt(unsigned char aes_iv[16], unsigned int keylen,
 }
 #endif /* MBEDTLS_AES_C */
 
+#if defined(MBEDTLS_DES_C) || defined(MBEDTLS_AES_C)
+static int pem_check_pkcs_padding(unsigned char *input, size_t input_len, size_t *data_len)
+{
+    size_t pad_len = input[input_len - 1];
```

Comment:
I think that at this point, there's no guarantee that `input_len != 0`. This can happen if the base64 data is empty (after removing whitespace), on systems where `malloc(0)` returns non-null (which includes Linux and FreeBSD). The decryption code guarantees that the input is a whole number of blocks, but that number can be 0.

Please add a test case with encryption and an empty ciphertext, and a `len=0` check if necessary.

Reply:
Mmmm quite surprising: I added the test you mentioned with empty ciphertext and I was expecting this assignment to return something like "you are trying to access data which is way out of program's memory boundary", but that didn't happened on my PC. I mean `input_len` is 0 in this case, so `-1` should be like > 4GB (unless `size_t` is 64 bit, in which case it could even be worse that that).
Am I missing something here?

ps = of course I'm adding the extra check here as you suggested. My question was mostly for my curiosity

Reply:
wait, another question: does it make sense for PEM to have `len = 0` in general (i.e. ciphered or not)?

Reply:
`len=0` is not valid data, but at this point I think nothing has rejected it yet.

`input[-1]` is a buffer underflow by one byte, so it accesses whatever is before `input` on the stack/heap. Asan would complain.

Reply:
Oh, ok, that's what I was missing. It never happened to me to access the `-1` item of an array so I thought it wasn't allowed. Thanks!

Reply:
It's not allowed (except if you have a sub-array of a larger object). Buffer under/overflows are not allowed, but they aren't detected in general, which is why they're a security hole.